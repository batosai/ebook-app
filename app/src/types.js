
export const BOOKS_REQUEST = 'books:request';
export const BOOKS_SUCCESS = 'books:success';
export const BOOKS_FAILURE = 'books:failure';

export const BOOK_DELETE_REQUEST = 'book:delete';
export const BOOK_DELETE_SUCCESS = 'book:delete:success';

export const ASIDE_TOGGLE = 'aside:toggle';

export const COLLECTIONS_REQUEST = 'collections:request';
export const COLLECTIONS_SUCCESS = 'collections:success';
export const COLLECTIONS_FAILURE = 'collections:failure';
export const COLLECTIONS_FILTER  = 'collections:filter';

export const COLLECTION_ADD_REQUEST    = 'collection:add:request';
export const COLLECTION_EDIT_REQUEST    = 'collection:edit:request';
export const COLLECTION_DELETE_REQUEST = 'collection:delete:request';

export const LIBRARIES_REQUEST = 'libraries:request';
export const LIBRARIES_SUCCESS = 'libraries:success';
export const LIBRARIES_FAILURE = 'libraries:failure';

export const LIBRARY_ADD_REQUEST    = 'library:add:request';
export const LIBRARY_EDIT_REQUEST    = 'library:edit:request';
export const LIBRARY_DELETE_REQUEST = 'library:delete:request';
