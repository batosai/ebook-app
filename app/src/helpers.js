export function createReducer (initialState, reducers) {
  return (state = initialState, { type, payload }) => {
    const reducer = reducers[type];
    return reducer ? reducer(state, payload) : state;
  };
};

export function socketIoMiddleware(socket, channelName = "action") { return store => {
  socket.on(channelName, store.dispatch);

  return next => action => {
    if (action.emit) {
      socket.emit(channelName, action);
    }
    return next(action);
  }
}};
