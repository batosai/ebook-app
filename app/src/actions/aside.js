import { createAction } from '../helpers';
import * as types from '../types';

const toggle = createAction(types.ASIDE_TOGGLE, () => null);
export const asideToggle = text => dispatch => {
  dispatch(toggle(text));
};
