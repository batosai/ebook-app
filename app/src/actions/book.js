import { createAction } from '../helpers';
import * as types from '../types';

const bookRequest = createAction(types.BOOK_REQUEST, (id) => ({ id }), {emit:true});
export const findBookById = id => dispatch => {
  dispatch(bookRequest(id));
};

const bookDeleteRequest = createAction(types.BOOK_DELETE_REQUEST, (id) => ({ id }), {emit:true});
export const deleteBook = book => dispatch => {
  dispatch(bookDeleteRequest(book.id));
};

const bookEditRequest = createAction(types.BOOK_EDIT_REQUEST, (book) => ({book}), {emit:true});
export const editBook = book => dispatch => {
  dispatch(bookEditRequest(book));
};
