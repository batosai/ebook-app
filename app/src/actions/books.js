import { createAction } from '../helpers';
import { catchBook, catchBooksByCollection, modifyBook } from '../data/socket';

export const BOOKS_REQUEST = 'books:request';
export const BOOKS_SUCCESS = 'books:success';
export const BOOKS_FAILURE = 'books:failure';

const booksRequest = createAction(BOOKS_REQUEST, () => null);
const booksSuccess = createAction(BOOKS_SUCCESS, (items) => ({ items }));
const booksFailure = createAction(BOOKS_FAILURE, (error) => ({ error }));
export const getBooks = slug => dispatch => {
  dispatch(booksRequest(slug));
  catchBooksByCollection(slug)
  .then(books => dispatch(booksSuccess(books)))
  .catch(error => dispatch(booksFailure(error)));
};

export const BOOK_REQUEST = 'book:request';
export const BOOK_SUCCESS = 'book:success';
export const BOOK_FAILURE = 'book:failure';

const bookRequest = createAction(BOOK_REQUEST, (id) => ({ id }));
const bookSuccess = createAction(BOOK_SUCCESS, (book) => ({ book }));
const bookFailure = createAction(BOOK_FAILURE, (error) => ({ error }));
export const getBook = text => dispatch => {
  dispatch(bookRequest(text));
  catchBook(text)
  .then(book => dispatch(bookSuccess(book)))
  .catch(error => dispatch(bookFailure(error)));
};

export const BOOK_UPDATE  = 'book:update';
export const BOOK_UPDATE_FAILURE = 'list:update:failure';

const bookUpdateSuccess = createAction(BOOK_UPDATE, (book) => ({ book }));
const bookUpdateFailure = createAction(BOOK_UPDATE_FAILURE, (error) => ({ error }));

export const updateBook = (book, success) => dispatch => {
  modifyBook(book)
  .then((book) => {
    dispatch(bookUpdateSuccess(book));
    success();
  })
  .catch(error => dispatch(bookUpdateFailure(error)));
};
