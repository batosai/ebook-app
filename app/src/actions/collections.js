import { createAction } from '../helpers';
import { collections } from '../data/socket';

export const COLLECTIONS_REQUEST = 'collections:request';
export const COLLECTIONS_SUCCESS = 'collections:success';
export const COLLECTIONS_FAILURE = 'collections:failure';
export const COLLECTIONS_FILTER  = 'collections:filter';

const collectionsRequest = createAction(COLLECTIONS_REQUEST, () => null);
const collectionsSuccess = createAction(COLLECTIONS_SUCCESS, (collections) => ({ collections }));
const collectionsFailure = createAction(COLLECTIONS_FAILURE, (error) => ({ error }));
export const getCollections = () => dispatch => {
  dispatch(collectionsRequest());
  collections()
  .then(collections => dispatch(collectionsSuccess(collections)))
  .catch(error => dispatch(collectionsFailure(error)));
};

const collectionByLibrary = createAction(COLLECTIONS_FILTER, (id) => (id));
export const getCollectionByLibrary = id => dispatch => {
  dispatch(collectionByLibrary(id));
};
