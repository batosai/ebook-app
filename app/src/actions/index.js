import { asideToggle } from './aside';
export { asideToggle };

import { findBooksByCollectionId } from './books';
export { findBooksByCollectionId };

import { findBookById, deleteBook, editBook } from './book';
export { findBookById, deleteBook, editBook };

import { findCollections, findCollectionByLibrary, createCollection, editCollection, deleteCollection } from './collections';
export { findCollections, findCollectionByLibrary, createCollection, editCollection, deleteCollection };

import { findLibraries, createLibrary, editLibrary, deleteLibrary } from './libraries';
export { findLibraries, createLibrary, editLibrary, deleteLibrary };
