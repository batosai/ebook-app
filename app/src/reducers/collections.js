import { COLLECTIONS_SUCCESS } from '../actions/collections';

import { createReducer } from '../helpers';

const initialState = [ /* {id, img, title, author} */ ];

const reducers = {

  [COLLECTIONS_SUCCESS]: (prevState, payload) =>
    payload.collections,

}

export default createReducer(initialState, reducers);
