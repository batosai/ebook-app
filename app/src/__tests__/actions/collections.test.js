import * as actions from '../../actions/collections'
import * as types from '../../types'
import { createAction } from '../../helpers';

describe('Action', () => {
  it('to request collections', () => {
    const expectedAction = createAction(types.COLLECTIONS_REQUEST, () => null, {emit:true});

    expect(actions.collectionsRequest()).toEqual(expectedAction());
    expect(actions.collectionsRequest()).toMatchSnapshot();
  });

  it('to filter collections', () => {
    const expectedAction = createAction(types.COLLECTIONS_FILTER, (id) => (id));

    expect(actions.collectionByLibrary()).toEqual(expectedAction());
    expect(actions.collectionByLibrary()).toMatchSnapshot();
  });

  it('to add collections', () => {
    const expectedAction = createAction(types.COLLECTION_ADD_REQUEST, (data) => ({...data}), {emit:true});

    expect(actions.collectionAddRequest()).toEqual(expectedAction());
    expect(actions.collectionAddRequest()).toMatchSnapshot();
  });

  it('to edit collections', () => {
    const expectedAction = createAction(types.COLLECTION_EDIT_REQUEST, (data) => ({...data}), {emit:true});

    expect(actions.collectionEditRequest()).toEqual(expectedAction());
    expect(actions.collectionEditRequest()).toMatchSnapshot();
  });

  it('to delete collections', () => {
    const expectedAction = createAction(types.COLLECTION_DELETE_REQUEST, (id) => ({id}), {emit:true});

    expect(actions.collectionDeleteRequest()).toEqual(expectedAction());
    expect(actions.collectionDeleteRequest()).toMatchSnapshot();
  });
});
