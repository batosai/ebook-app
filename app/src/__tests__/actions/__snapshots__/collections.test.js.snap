exports[`Action to add collections 1`] = `
Object {
  "emit": true,
  "payload": Object {
    "library_id": 2,
    "title": "My collection",
  },
  "type": "collection:add:request",
}
`;

exports[`Action to delete collections 1`] = `
Object {
  "emit": true,
  "payload": Object {
    "id": 1,
  },
  "type": "collection:delete:request",
}
`;

exports[`Action to edit collections 1`] = `
Object {
  "emit": true,
  "payload": Object {
    "id": 1,
    "library_id": 2,
    "title": "My collection",
  },
  "type": "collection:edit:request",
}
`;

exports[`Action to filter collections 1`] = `
Object {
  "payload": 1,
  "type": "collections:filter",
}
`;

exports[`Action to request collections 1`] = `
Object {
  "emit": true,
  "type": "collections:request",
}
`;
